File: backend/app/api/members.py
from fastapi import APIRouter
from app.schemas.member import MemberCreate, MemberResponse
from app.services import member_service

router = APIRouter(prefix="/members", tags=["members"])


@router.get("/", response_model=list[MemberResponse])
def list_members():
    return member_service.get_all_members()


@router.get("/{member_id}", response_model=MemberResponse)
def get_member(member_id: str):
    return member_service.get_member_by_id(member_id)


@router.post("/", response_model=MemberResponse, status_code=201)
def create_member(payload: MemberCreate):
    return member_service.create_member(payload)

File: backend/src/api/members.ts
import { Router } from "express";
import { MemberCreateSchema, MemberUpdateSchema, MemberUpdate } from "../schemas/member";
import { MemberService } from "../services/memberService";
import { pickPresentKeys } from "../middleware/validateBody";

const MEMBER_UPDATE_KEYS = Object.keys(MemberUpdateSchema.shape) as (keyof MemberUpdate)[];

export function createMembersRouter(deps: { memberService: MemberService }): Router {
  const router = Router();
  const { memberService } = deps;

  router.get("/", async (_req, res, next) => {
    try {
      res.json(await memberService.getAllMembers());
    } catch (e) {
      next(e);
    }
  });

  router.get("/:id", async (req, res, next) => {
    try {
      res.json(await memberService.getMemberById(req.params.id));
    } catch (e) {
      next(e);
    }
  });

  router.post("/", async (req, res, next) => {
    try {
      const payload = MemberCreateSchema.parse(req.body);
      res.status(201).json(await memberService.createMember(payload));
    } catch (e) {
      next(e);
    }
  });

  router.patch("/:id", async (req, res, next) => {
    try {
      const parsed = MemberUpdateSchema.parse(req.body);
      const patch = pickPresentKeys(req.body, parsed, MEMBER_UPDATE_KEYS);
      res.json(await memberService.updateMember(req.params.id, patch));
    } catch (e) {
      next(e);
    }
  });

  router.delete("/:id", async (req, res, next) => {
    try {
      await memberService.deleteMember(req.params.id);
      res.status(204).send();
    } catch (e) {
      next(e);
    }
  });

  return router;
}
